gensim/models/base_any2vec.py

```diff
@@ -0,0 +1,792 @@
+        """Fill jobs queue using the input `sentences` iterator."""
+        job_batch, batch_size = [], 0
+        pushed_words, pushed_examples = 0, 0
+        next_job_params = self._get_job_params(cur_epoch)
+        job_no = 0
+
+        for data_idx, data in enumerate(data_iterator):
+            data_length = self._raw_word_count([data])
+
+            # can we fit this sentence into the existing job batch?
+            if batch_size + data_length <= self.batch_words:
+                # yes => add it to the current job
+                job_batch.append(data)
+                batch_size += data_length
+            else:
+                job_no += 1
+                job_queue.put((job_batch, next_job_params))
+
+                # update the learning rate for the next job
+                if total_examples:
+                    # examples-based decay
+                    pushed_examples += len(job_batch)
+                    epoch_progress = 1.0 * pushed_examples / total_examples
+                else:
+                    # words-based decay
+                    pushed_words += self._raw_word_count(job_batch)
+                    epoch_progress = 1.0 * pushed_words / total_words
+                next_job_params = self._update_job_params(next_job_params, epoch_progress, cur_epoch)
+
+                # add the sentence that didn't fit as the first item of a new job
+                job_batch, batch_size = [data], data_length
+        # add the last job too (may be significantly smaller than batch_words)
+        if job_batch:
+            job_no += 1
+            job_queue.put((job_batch, next_job_params))
+
+        if job_no == 0 and self.train_count == 0:
+            logger.warning(
+                "train() called with an empty iterator (if not intended, "
+                "be sure to provide a corpus that offers restartable iteration = an iterable)."
+            )
+
+        # give the workers heads up that they can finish -- no more work!
+        for _ in xrange(self.workers):
+            job_queue.put(None)
+        logger.debug("job loop exiting, total %i jobs", job_no)
+
+    def _log_progress(self, job_queue, progress_queue, cur_epoch, example_count, total_examples,
+                      raw_word_count, total_words, trained_word_count, elapsed):
+        raise NotImplementedError
+
+    def _log_epoch_end(self, cur_epoch, example_count, total_examples, raw_word_count, total_words,
+                       trained_word_count, elapsed, job_tally):
+        raise NotImplementedError
+
+    def _log_train_end(self, raw_word_count, trained_word_count, total_elapsed):
+        raise NotImplementedError
+
+    def _log_epoch_progress(self, progress_queue, job_queue, cur_epoch=0, total_examples=None, total_words=None,
+                            report_delay=1.0):
+        example_count, trained_word_count, raw_word_count = 0, 0, 0
+        start, next_report = default_timer() - 0.00001, 1.0
```

Comment:
I forgot, why this `- 0.00001` needed?

Reply:
It is a guard against next sample being identical. Look [here](https://github.com/RaRe-Technologies/gensim/blob/5ea3eb2c37cf57665d55c98168ff5b343c479e12/gensim/models/doc2vec.py#L692).

Reply:
Would it be feasible to make this a callback? Maybe `EpochProgressLoggingCallback` and use it as the default?